render/drm_dumb_allocator.c

```diff
@@ -0,0 +1,212 @@
+	create.bpp = 32;
+
+	if (drmIoctl(alloc->drm_fd, DRM_IOCTL_MODE_CREATE_DUMB, &create) != 0) {
+		wlr_log(WLR_ERROR, "failed to create drm dumb buffer: %s",
+				strerror(errno));
+		goto create_err;
+	}
+
+	buffer->width = create.width;
+	buffer->height = create.height;
+
+	buffer->stride = create.pitch;
+	buffer->handle = create.handle;
+
+	buffer->drm_fd = alloc->drm_fd;
+
+	struct drm_mode_map_dumb map = {0};
+	map.handle = buffer->handle;
+
+	if (drmIoctl(alloc->drm_fd, DRM_IOCTL_MODE_MAP_DUMB, &map) != 0) {
+		wlr_log(WLR_ERROR, "failed to map drm dumb buffer: %s",
+				strerror(errno));
+		goto create_destroy;
+	}
+
+	buffer->data = mmap(NULL, create.size, PROT_WRITE, MAP_SHARED,
+			alloc->drm_fd, map.offset);
+	if (buffer->data == MAP_FAILED) {
+		wlr_log(WLR_ERROR, "failed to mmap drm dumb buffer: %s",
+				strerror(errno));
+		goto create_destroy;
+	}
+
+	buffer->size = create.size;
+
+	memset(buffer->data, 0, create.size);
+
+	int prime_fd;
+	if (drmPrimeHandleToFD(alloc->drm_fd, buffer->handle, DRM_CLOEXEC,
+			&prime_fd) != 0) {
+		wlr_log(WLR_ERROR, "failed to get prime handle from GEM: %s",
+				strerror(errno));
+		goto create_destroy;
+	}
+
+	buffer->dmabuf = (struct wlr_dmabuf_attributes){
+		.width = buffer->width,
+		.height = buffer->height,
+		.format = format->format,
+		.n_planes = 1,
+		.offset[0] = 0,
+		.stride[0] = buffer->stride,
+		.fd[0] = prime_fd,
+	};
+
+	wlr_log(WLR_DEBUG, "Allocated %u x %u DRM dumb buffer (GEM %" PRIu32 ")",
+			buffer->width, buffer->height, buffer->handle);
+
+	return buffer;
+create_destroy:
+	finish_buffer(buffer);
+create_err:
+	free(buffer);
+	return NULL;
+}
+
+static bool buffer_get_data_ptr(struct wlr_buffer *wlr_buffer, void **data,
+		size_t *stride) {
+	struct wlr_drm_dumb_buffer *buf = drm_dumb_buffer_from_buffer(wlr_buffer);
+	*data = buf->data;
+	*stride = buf->stride;
+	return true;
+}
+
+static bool buffer_get_dmabuf(struct wlr_buffer *wlr_buffer,
+		struct wlr_dmabuf_attributes *attribs) {
+	struct wlr_drm_dumb_buffer *buf = drm_dumb_buffer_from_buffer(wlr_buffer);
+	memcpy(attribs, &buf->dmabuf, sizeof(buf->dmabuf));
+	return true;
+}
+
+static void buffer_destroy(struct wlr_buffer *wlr_buffer) {
+	struct wlr_drm_dumb_buffer *buf = drm_dumb_buffer_from_buffer(wlr_buffer);
+	finish_buffer(buf);
+	free(buf);
+}
+
+static const struct wlr_buffer_impl buffer_impl = {
+	.destroy = buffer_destroy,
+	.get_dmabuf = buffer_get_dmabuf,
+	.get_data_ptr = buffer_get_data_ptr,
+};
+
+static const struct wlr_allocator_interface allocator_impl;
+
+static struct wlr_drm_dumb_allocator *drm_dumb_alloc_from_alloc(
+		struct wlr_allocator *wlr_alloc) {
+	assert(wlr_alloc->impl == &allocator_impl);
+	return (struct wlr_drm_dumb_allocator *)wlr_alloc;
+}
+
+static struct wlr_buffer *allocator_create_buffer(
+		struct wlr_allocator *wlr_alloc, int width, int height,
+		const struct wlr_drm_format *drm_format) {
+	if (drm_format->format != DRM_FORMAT_ARGB8888
```

Comment:
Maybe we can move this check in `create_buffer`? It would be closed to the assumption `bpp = 32`.

Reply:
Yup, this will have to change one #2779 is merged.